x/concentrated-liquidity/fees_test.go

```diff
@@ -448,3 +449,221 @@ func (suite *KeeperTestSuite) TestChargeFee() {
+
+		// inputs parameters.
+		owner     sdk.AccAddress
+		lowerTick int64
+		upperTick int64
+
+		// expectations.
+		expectedFeesClaimed sdk.Coins
+		expectedError       error
+	}{
+		// imagine single swap over entire position
+		// crossing left > right and stopping above upper tick
+		// In this case, only the upper tick accumulator must have
+		// been updated when crossed.
+		// Since we track fees accrued below a tick, upper tick is updated
+		// while lower tick is zero.
+		"single swap left -> right: 2 ticks, one share, current tick > upper tick": {
+			initialLiquidity: sdk.OneDec(),
+
+			lowerTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			globalFeeGrowth: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			owner:     ownerWithValidPosition,
+			lowerTick: 0,
+			upperTick: 1,
+
+			currentTick: 2,
+
+			expectedFeesClaimed: sdk.NewCoins(sdk.NewCoin(ETH, sdk.NewInt(10))),
+		},
+		"single swap left -> right: 3 ticks, two shares, current tick > upper tick": {
+			initialLiquidity: sdk.NewDec(2),
+
+			lowerTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			globalFeeGrowth: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			owner:     ownerWithValidPosition,
+			lowerTick: 0,
+			upperTick: 2,
+
+			currentTick: 3,
+
+			expectedFeesClaimed: sdk.NewCoins(sdk.NewCoin(ETH, sdk.NewInt(20))),
+		},
+		"single swap left -> right: 2 ticks, one share, current tick == upper tick": {
+			initialLiquidity: sdk.OneDec(),
+
+			lowerTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			globalFeeGrowth: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			owner:     ownerWithValidPosition,
+			lowerTick: 0,
+			upperTick: 1,
+
+			currentTick: 1,
+
+			expectedFeesClaimed: sdk.NewCoins(sdk.NewCoin(ETH, sdk.NewInt(10))),
+		},
+
+		// imagine single swap over entire position
+		// crossing right -> left and stopping at lower tick
+		// In this case, all fees must have been accrued inside the tick
+		// Since we track fees accrued below a tick, both upper and lower position
+		// ticks are zero
+		"single swap right -> left: 2 ticks, one share, current tick == lower tick": {
+			initialLiquidity: sdk.OneDec(),
+
+			// lower tick accumulator is not updated yet.
+			lowerTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+
+			globalFeeGrowth: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			owner:     ownerWithValidPosition,
+			lowerTick: 0,
+			upperTick: 1,
+
+			currentTick: 0,
+
+			expectedFeesClaimed: sdk.NewCoins(sdk.NewCoin(ETH, sdk.NewInt(10))),
+		},
```

Comment:
Can we add a test case where the feeGrowthOutside is equal to the globalFeeGrowth so we expect zero fees to claim?

Reply:
done